lib/devicetrust/assert/assert_test.go

```diff
@@ -0,0 +1,164 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package assert_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/lib/devicetrust/assert"
+	"github.com/gravitational/teleport/lib/devicetrust/authn"
+	"github.com/gravitational/teleport/lib/devicetrust/enroll"
+	"github.com/gravitational/teleport/lib/devicetrust/testenv"
+)
+
+func TestCeremony(t *testing.T) {
+	t.Parallel()
+
+	env := testenv.MustNew(
+		testenv.WithAutoCreateDevice(true),
+	)
+
+	devicesClient := env.DevicesClient
+	ctx := context.Background()
+
+	macDev, err := testenv.NewFakeMacOSDevice()
+	require.NoError(t, err, "NewFakeMacOSDevice errored")
+
+	linuxDev := testenv.NewFakeLinuxDevice()
+
+	tests := []struct {
+		name string
+		dev  testenv.FakeDevice
+	}{
+		{
+			name: "ok (macOs)",
+			dev:  macDev,
+		},
+		{
+			name: "ok (TPM)",
+			dev:  linuxDev,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Enroll the device before we attempt authentication (auto creates device
+			// as part of ceremony)
+			dev := test.dev
+			enrollC := enroll.Ceremony{
+				GetDeviceOSType:         dev.GetDeviceOSType,
+				EnrollDeviceInit:        dev.EnrollDeviceInit,
+				SignChallenge:           dev.SignChallenge,
+				SolveTPMEnrollChallenge: dev.SolveTPMEnrollChallenge,
+			}
+			_, err := enrollC.Run(ctx, devicesClient, false, testenv.FakeEnrollmentToken)
+			require.NoError(t, err, "EnrollDevice errored")
+
+			assertC, err := assert.NewCeremony(assert.WithNewAuthnCeremonyFunc(func() *authn.Ceremony {
+				return &authn.Ceremony{
+					GetDeviceCredential: func() (*devicepb.DeviceCredential, error) {
+						return dev.GetDeviceCredential(), nil
+					},
+					CollectDeviceData:            dev.CollectDeviceData,
+					SignChallenge:                dev.SignChallenge,
+					SolveTPMAuthnDeviceChallenge: dev.SolveTPMAuthnDeviceChallenge,
+					GetDeviceOSType:              dev.GetDeviceOSType,
+				}
+			}))
+			require.NoError(t, err, "NewCeremony errored")
+
+			authnStream, err := devicesClient.AuthenticateDevice(ctx)
+			require.NoError(t, err, "AuthenticateDevice errored")
+
+			// Typically this would be some other, non-DeviceTrustService stream, but
+			// here this is a good way to test it (as it runs actual fake device
+			// authn)
+			if err := assertC.Run(ctx, &assertStreamAdapter{
+				stream: authnStream,
+			}); err != nil {
+				t.Errorf("Run returned err=%q, want nil", err)
+			}
+		})
+	}
+}
+
+type assertStreamAdapter struct {
+	stream devicepb.DeviceTrustService_AuthenticateDeviceClient
+}
+
+func (s *assertStreamAdapter) Recv() (*devicepb.AssertDeviceResponse, error) {
```

Comment:
Can we use `authnStreamAdapter` ?
It seems we are trying to test it but instead duplicate almost the exact code and test that instead.

Reply:
This one is the opposite direction of authnStreamAdapter, although it's rather similar.

authnStreamAdapter is a private type and I typically don't like reaching into privates for testing. This would break the test if I ever refactor the prod code to stop "adapting" streams (which I actually might, if time permits).